src/vipw.c

```diff
@@ -243,13 +243,13 @@ vipwedit (const char *file, int (*file_lock) (void), int (*file_unlock) (void))
 	/* if SE Linux is enabled then set the context of all new files
 	   to be the context of the file we are editing */
 	if (is_selinux_enabled () != 0) {
-		security_context_t passwd_context=NULL;
+		char *passwd_context_raw = NULL;
 		int ret = 0;
-		if (getfilecon (file, &passwd_context) < 0) {
+		if (getfilecon_raw (file, &passwd_context_raw) < 0) {
 			vipwexit (_("Couldn't get file context"), errno, 1);
 		}
-		ret = setfscreatecon (passwd_context);
-		freecon (passwd_context);
+		ret = setfscreatecon_raw (passwd_context_raw);
```

Comment:
Why do  _raw here?  The setfscreatecon manpage only says that it does "not perform context translation" but I don't know what that's referring to.

Note - I'm not doubting that it's the right thing, I just want to understand why it it.

Reply:
A SELinux context consists of 3 or 4 parts (separated by `:`), depending on whether the loaded policy enables MCS/MLS; for example `system_u:object_r:user_t` or `system_u:object_r:user_t:s0-s0:c0.c1023`.
The last part, in this case `s0-s0:c0.c1023`, is the sensitivity and/or category part of the context.

To improve usability it is possible to set an alias name for those `sX`/`cX` identifiers.
libselinux supplies the functions `selinux_trans_to_raw_context()` and `selinux_raw_to_trans_context()`, which are called by the non-raw functions behind the scenes, to convert raw contexts into human readable ones and vice versa.

It is useful to use the translated non-raw functions if the context is either being presented to the user, e.g. by ls(1) or ps(1), or the context is read from a user supplied input, e.g. pam_selinux(7).

shadow does neither of these two things, so translating the context is not necessary.

See also https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/5/html/deployment_guide/sec-mcs-config-categories, or [setrans.conf(5)](https://manpages.debian.org/unstable/mcstrans/setrans.conf.5.en.html)